File: example/android/app/src/main/kotlin/br/dig/dev/flutter/flutter_reference_example/ExamplePlugin.kt
package br.dig.dev.flutter.flutter_reference_example

import br.dig.dev.flutter.flutter_reference.References
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.security.SecureRandom

class ExamplePlugin : FlutterPlugin, MethodChannel.MethodCallHandler {

    private lateinit var channel: MethodChannel

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "example_plugin")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        return when (call.method) {
            "create" -> {
                val instance = when (SecureRandom().nextInt(4)) {
                    0 -> Apple
                    1 -> Banana
                    2 -> Orange
                    3 -> NotFruit("NotFruit")
                    else -> NotFruit("Error")
                }
                result.success(References.add(instance))
            }

            "name" -> {
                val instance = References.get<ExampleClass>(call.arguments<String>()!!)
                result.success(instance.name)
            }

            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

}

File: example/ios/Runner/ExamplePlugin.swift
import flutter_reference

class ExamplePlugin: NSObject, FlutterPlugin {
  static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "example_plugin", binaryMessenger: registrar.messenger())
    let instance = ExamplePlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "create":
      let instance: ExampleClass
      switch Int.random(in: 0..<4) {
      case 0: instance = Apple()
      case 1: instance = Banana()
      case 2: instance = Orange()
      case 3: instance = NotFruit("NotFruit")
      default: instance = NotFruit("Error")
      }
      result(References.add(instance))

    case "name":
      let instance: ExampleClass = References.get(call.arguments as! String)
      result(instance.name)

    default:
      result(FlutterMethodNotImplemented)
    }
  }
}
